### PR TITLE
Rename xtu to ctu

### PR DESCRIPTION
**Pull request by gamesh411**
This request contains 2 commits. One for changing the remaining parameters, variables etc. that refer to the cross-translation-unit feature as "xtu" to "ctu". The other fixes the parameter passing and ctu-dir layout used by the ctu-build-new scripts, so it is up to date with current clang.

### CODE REVIEW

File: test/Analysis/ctu-main_usr.c [hunk 1]
```diff
@@ -1,7 +1,7 @@
-// RUN: mkdir -p %T/xtudir2
-// RUN: %clang_cc1 -triple x86_64-pc-linux-gnu -emit-pch -o %T/xtudir2/xtu-other.c.ast %S/Inputs/xtu-other.c
-// RUN: cp %S/Inputs/externalFnMap2_usr.txt %T/xtudir2/externalFnMap.txt
-// RUN: %clang_cc1 -triple x86_64-pc-linux-gnu -fsyntax-only -std=c89 -analyze -analyzer-checker=core,debug.ExprInspection -analyzer-config xtu-dir=%T/xtudir2 -analyzer-config use-usr=true -analyzer-config reanalyze-xtu-visited=true -verify %s
+// RUN: mkdir -p %T/ctudir2
+// RUN: %clang_cc1 -triple x86_64-pc-linux-gnu -emit-pch -o %T/ctudir2/ctu-other.c.ast %S/Inputs/ctu-other.c
+// RUN: cp %S/Inputs/externalFnMap2_usr.txt %T/ctudir2/externalFnMap.txt
+// RUN: %clang_cc1 -triple x86_64-pc-linux-gnu -fsyntax-only -std=c89 -analyze -analyzer-checker=core,debug.ExprInspection -analyzer-config ctu-dir=%T/ctudir2 -analyzer-config use-usr=true -analyzer-config reanalyze-ctu-visited=true -verify %s
```
Comment (Xazax-hun) [inline test/Analysis/ctu-main_usr.c]: `-analyzer-config reanalyze-ctu-visited=true` part is no longer needed, You could remove this from the runline.

File: test/Analysis/ctu-main_usr.cpp [hunk 2]
```diff
@@ -1,12 +1,12 @@
...
-// RUN: %clang_cc1 -triple x86_64-pc-linux-gnu -fsyntax-only -analyze -analyzer-checker=core,debug.ExprInspection -analyzer-config xtu-dir=%T/xtudir -analyzer-config use-usr=true -analyzer-config reanalyze-xtu-visited=true -std=c++11 -analyzer-display-ctu-progress 2>&1 %s | FileCheck %s
-
-// CHECK: ANALYZE (CTU loaded AST for source file): {{.*}}/xtu-other.cpp
-// CHECK: ANALYZE (CTU loaded AST for source file): {{.*}}/xtu-chain.cpp
+// RUN: mkdir -p %T/ctudir
+// RUN: %clang_cc1 -triple x86_64-pc-linux-gnu -std=c++11 -emit-pch -o %T/ctudir/ctu-other.cpp.ast %S/Inputs/ctu-other.cpp
+// RUN: %clang_cc1 -triple x86_64-pc-linux-gnu -std=c++11 -emit-pch -o %T/ctudir/ctu-chain.cpp.ast %S/Inputs/ctu-chain.cpp
+// RUN: cp %S/Inputs/externalFnMap_usr.txt %T/ctudir/externalFnMap.txt
+// RUN: %clang_cc1 -triple x86_64-pc-linux-gnu -fsyntax-only -analyze -analyzer-checker=core,debug.ExprInspection -analyzer-config ctu-dir=%T/ctudir -analyzer-config use-usr=true -analyzer-config reanalyze-ctu-visited=true -std=c++11 -verify %s
```
Comment (Xazax-hun) [inline test/Analysis/ctu-main_usr.cpp]: Same as above, command line could be simplified. Similarily we only support USR now, so `-analyzer-config use-usr=true` can be removed as well. It looks like I forgot to update the tests after those changes.

File: tools/ctu-build-new/ctu-analyze.py [hunk 3]
```diff
@@ -217,6 +218,22 @@ def get_compiler_and_arguments(cmd):
...
+    arch = ""
+    clang_cmd = []
+    clang_cmd.append(os.path.join(clang_path, 'clang'))
+    clang_cmd.append("-###")
+    clang_cmd.extend(clang_args)
+    clang_cmd.append(source)
+    clang_out = subprocess.check_output(clang_cmd, stderr=subprocess.STDOUT, shell=False)
+    clang_params=shlex.split(clang_out)
+    i=0
+    while i<len(clang_params) and clang_params[i]!="-triple":
```
Comment (Xazax-hun) [inline tools/ctu-build-new/ctu-analyze.py]: Use spaces around operators.

File: tools/ctu-build-new/ctu-build.py [hunk 4]
```diff
@@ -221,12 +237,22 @@ def create_external_fn_maps(ctuindir):
...
+    triple_paths = glob.glob(os.path.join(ctuindir, TEMP_EXTERNAL_FNMAP_FOLDER, '*'))
+    triple_dirs = filter(os.path.isdir, triple_paths)
+    triples = map(os.path.basename, triple_dirs)
+    for triple in triples:
+        create_external_fn_maps_for_triple(ctuindir, triple)
+
+
+if not os.path.exists(mainargs.ctuindir):
+    os.makedirs(mainargs.ctuindir)
+clear_file(os.path.join(mainargs.ctuindir, 'cfg.txt'))
```
Comment (Xazax-hun) [inline tools/ctu-build-new/ctu-build.py]: We no longer generate `cfg.txt`. I think the related code can be removed. Maybe there are other files we no longer generate from these ones.

File: tools/ctu-ci/run_all.sh [hunk 5]
```diff
@@ -24,7 +24,7 @@ cd projects
 rm statistics.txt
 rm detailed_stats.txt
 
-export PATH=/mnt/storage/xtu-service/clang_build/buildrelwdeb/bin:/mnt/storage/xtu-service/clang_build/codechecker/build/CodeChecker/bin:$PATH
+export PATH=/mnt/storage/ctu-service/clang_build/buildrelwdeb/bin:/mnt/storage/ctu-service/clang_build/codechecker/build/CodeChecker/bin:$PATH
```
Comment (Xazax-hun) [inline tools/ctu-ci/run_all.sh]: These paths are specific to the environment we use for testing. They should not be rewritten.

File: tools/ctu-ci/analyze_project.sh [hunk 6]
```diff
@@ -46,76 +46,76 @@ do
...
+cp -rf .ctu-out-ctu ./"$PROJNAME"_ctu-out-ctu
+rm -rf .ctu-out-ctu/
+#INITIAL CTU RUN
+#/usr/bin/time -f "%e" ../../../clang_build/clang/tools/ctu-build-new/ctu-build.py $USR $REPARSE -b buildlog.json --clang-path /mnt/storage/ctu-service/clang_build/buildrelwdeb/bin/ -v -j 16 2>"$PROJNAME"_buildtime.out | tee "$PROJNAME"_build.out
+#../../../clang_build/clang/tools/ctu-build-new/ctu-analyze.py $USR $REPARSE -b buildlog.json -o .ctu-out-init -j 16 -v --clang-path /mnt/storage/ctu-service/clang_build/buildrelwdeb/bin/ --analyze-cc-path /mnt/storage/ctu-service/clang_build/clang/tools/scan-build-py/bin/ --log-passed-build passed_buildlog.json | tee "$PROJNAME"_init.out
+#NOCTU-RUN
+#../../../clang_build/clang/tools/ctu-build-new/ctu-analyze.py $MEMPROF --record-coverage -b passed_buildlog.json -o .ctu-out-noctu -j 16 -v --clang-path /mnt/storage/ctu-service/clang_build/buildrelwdeb/bin/ --analyze-cc-path /mnt/storage/ctu-service/clang_build/clang/tools/scan-build-py/bin/ --no-ctu | tee "$PROJNAME"_noCTU.out
+PATH=/mnt/storage/ctu-service/clang_build/buildrelwdeb/bin/:$PATH $CC analyze -o .ctu-out-noctu --analyzers clangsa -j 16 buildlog.json | tee "$PROJNAME"_noCTU.out
```
Comment (Xazax-hun) [inline tools/ctu-ci/analyze_project.sh]: Some of these absolute paths are environment specific, and should not be renamed.

**Comment by gamesh411** [comment]
Applied fixes, the ctu-ci tool should be reviewed, or at least looked at in order to ensure that it works as well.

File: test/Analysis/ctu-main_usr.cpp (hunk 2 above)
Comment (Xazax-hun) [reply test/Analysis/ctu-main_usr.cpp]: The redundant `-analyzer-config use-usr=true -analyzer-config reanalyze-ctu-visited=true` were not removed here and bellow.

File: tools/ctu-build-new/ctu-analyze.py [hunk 7]
```diff
@@ -149,21 +150,21 @@ def check_executable_available(exe_name, arg_path):
     analyzer_params += ['-analyzer-disable-checker', mainargs.disable_checkers]
 if mainargs.reparse:
     analyzer_params += ['-analyzer-config',
-                        'xtu-reparse='+os.path.abspath(mainargs.buildlog)]
-if not mainargs.no_xtu:
-    analyzer_params += ['-analyzer-config',
-                        'xtu-dir=' + os.path.abspath(mainargs.xtuindir)]
+                        'ctu-reparse='+os.path.abspath(mainargs.buildlog)]
+#if not mainargs.no_ctu:
```
Comment (Xazax-hun) [inline tools/ctu-build-new/ctu-analyze.py]: Why do you have this commented out code?

File: tools/ctu-build-new/ctu-analyze.py [hunk 8]
```diff
@@ -217,6 +218,22 @@ def get_compiler_and_arguments(cmd):
...
+    arch = ""
+    clang_cmd = []
+    clang_cmd.append(os.path.join(clang_path, 'clang'))
+    clang_cmd.append("-###")
+    clang_cmd.extend(clang_args)
+    clang_cmd.append(source)
+    clang_out = subprocess.check_output(clang_cmd, stderr=subprocess.STDOUT, shell=False)
+    clang_params=shlex.split(clang_out)
+    i = 0
+    while i<len(clang_params) and clang_params[i] != "-triple":
```
Comment (Xazax-hun) [inline tools/ctu-build-new/ctu-analyze.py]: Missing spaces around `<`.

**Comment by whisperity** [comment]
Is this PR still relevant? There are a lot of conflicts, but also CTU 1st version has been upstreamed.

**Comment by Xazax-hun** [comment]
With tools/ctu-ci being replaced by https://github.com/Xazax-hun/csa-testbench and ctu-clang6 patch is using ctu instead of xtu everywhere this pull request is no longer relevant.

**Pull request closed without merging**